### DISCUSSION TITLE
How to use variants with Box, Flex, etc.?

### DISCUSSION THREAD

**Discussion by TranquilMarmot** [system-ui/theme-ui#2011] [Q&A] [answered]
Looking at the docs here: https://theme-ui.com/components/variants

It says:

> The following components support the variant prop, but do not include default variant keys.
>    - `Box`
>    - `Flex`
>    - etc.

But it does not specify how to use `variant` with those components. I see with other components, there is a specific "Variant group" in the table (i.e. `Grid` goes with `grids`, `Button` goes with `buttons`) but there's no variant group for `Box` or `Flex`.

Using the `Theme` type to specify a theme in TypeScript, there is no way to add a `boxes` variant group. The `Box` component seems to be [looking at a `variants` key](https://github.com/system-ui/theme-ui/blob/8d80c4ea7139173872993faf096f322bd822742f/packages/components/src/Box.js#L19-L20) but you also cannot add that in TypeScript.

How do you define variants for `Box` and all of the other components that don't have their own variant group?

**Comment by lachlanjc** [comment, marked as answer]
Ah, we should totally make this clearer! If there's no theme key, that means you can use any variant from anywhere in the theme, but it needs a category prefix. E.g. you can write `<Grid variant="cards.sunken"` to make a Grid look like a Card, or `variant="layout.container"` to add Container's styling. (See the former [in use here](https://sourcegraph.com/github.com/hackclub/theme/-/blob/docs/pages/index.js?L166:10).)

**Reply by CamilleHbp** [reply]
Does it truly work ? I'm trying to use a Flex component with a variant from my theme but it does not pick it up.

```js
const exampleTheme = {
  ...
  layout: {
    column: {
      flexDirection: 'column',
    },
};
```

And I call the component like this: 
```jsx
<Flex variant="layout.column"> {children here...} </Flex>
```


Nothing happens. When calling it using the `sx` prop, everything works fine.


```jsx
<Flex sx={{ variant="layout.column" }}> {children here...} </Flex>
```

Am I miss-using  it? If yes, documentation would really need some clarification. ^^

**Reply by lachlanjc** [reply]
Funky! That should work either way. We’ve got a test for this functionality [on the Box component](https://github.com/system-ui/theme-ui/blob/develop/packages/components/test/Box.spec.tsx#L54), but not on Flex. Is it only happening on Flex or is the variant prop broken elsewhere?

**Reply by CamilleHbp** [reply]
Sorry, I'll try to check tomorrow and answer you! I keep missing those GitHub notifications. 😞

**Reply by CamilleHbp** [reply]
It does not seem to work with the Box component either. I'm wondering if I'm doing something wrong.

**Reply by lachlanjc** [reply]
Can you share your theme & the code you're having issues with?

**Reply by CamilleHbp** [reply]
Okay, I found out my issue. It had nothing to do with theme-ui. I had a problem in the build process of my library. I fixed it by having vite preserve the jsx import. Sorry for the inconvenience!
